instrumentation/http/src/main/java/brave/http/HttpRequestParser.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2013-2020 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package brave.http;
+
+import brave.SpanCustomizer;
+import brave.propagation.ExtraFieldPropagation;
+import brave.propagation.TraceContext;
+
+/**
+ * Use this to control the request data recorded for an {@link TraceContext#sampledLocal() sampled
+ * HTTP client or server span}.
+ *
+ * <p>Here's an example that changes the span name and records the HTTP url instead of the path.
+ * <pre>{@code
+ * httpTracing = httpTracing.toBuilder()
+ *   .clientRequestParser((request, context, span) -> {
+ *     span.name(spanName(adapter, request)); // default span name
+ *     span.tag("http.url", request.url()); // the whole url, not just the path
+ *   }).build();
+ * }</pre>
+ *
+ * <h3>Relationship to {@link HttpParser}</h3>
+ * This handles the request side of what {@link HttpParser} formerly did. This uses the {@link
+ * HttpRequest} type directly, avoiding the deprecated {@link HttpAdapter} type. By constraining
+ * this to only request data, it is implementable as a lambda or method reference. This also passes
+ * the trace context as a parameter, lowering parsing overhead vs relying on the current span.
+ *
+ * <h3>The {@link TraceContext} parameter</h3>
+ * The {@link TraceContext} parameter provides for advanced data handling, such as {@link
+ * ExtraFieldPropagation#get(TraceContext, String) extra fields}. This is explicitly passed as you
+ * cannot rely on {@link brave.Tracer#currentSpan()} being the same value.
+ *
+ * @see HttpResponseParser
+ */
+public interface HttpRequestParser {
+  HttpRequestParser DEFAULT = new HttpRequestParser.Default();
+
+  /**
+   * Implement to choose what data from the http request are parsed into the span representing it.
+   *
+   * @see Default
+   */
+  void parse(HttpRequest request, TraceContext context, SpanCustomizer span);
+
+  /**
+   * The default data policy sets the span name to the HTTP method and adds the "http.method" and
+   * "http.path" tags.
+   */
+  // Eventhough the default span name is the method, we have no way of knowing that a user hasn't
+  // overwritten the name to something else. If that occurs during response parsing, it is too late
```

Comment:
I think it's good to add the method as a tag either way, but for this statement, can't we pass the request to the response parser?

Reply:
https://github.com/openzipkin/brave/issues/1086